pull new: Error if no base branch can be inferred reliably

llucax opened the pull request:
If there's no tracking branch configured and no `hub.pullbase` set, the default base branch for `pull new` is `master`. This is good default most of the time. However, if there is no local branch named `master`, this is almost certainly wrong choice. I'd like git-hub to require user to explicitly use `-b`/`--base` in such case.

**Note:** This comment was written by @jwilk, GitHub will change the issue author when is converted to a PR by attaching code :unamused:

joshtriplett commented:
This has become much more common; many projects now use a different default branch, such as "main".

It'd be helpful if git-hub defaulted to whatever the upstream default branch is.

llucax commented:
This is indeed more important now that `master` is even less omnipresent as is used to be, I increased the priority to reflect that. I'm making no promise but I'll try to give it a look in the near future. PRs are more than welcome.

PS: I also marked it as a breaking change, as it will change silently for users relying on this default. I guess there should be a way to opt-in for this behaviour or some other way to provide a non-breaking path to the upgrade so this doesn't have to wait for another major release.

llucax committed:
pull: Remove the last resort default for --base

Both `pull new` and `pull attach` have a last resort `master` default
for `--base` but it makes little sense, as repositories sometimes have
a different default branch, or have no `master` at all. Now that GitHub
have changed the default branch to `main` for new projects (and many
projects are moving away from using `master` as a name for anything
altogether) it makes less sense that ever to use this last resort
default.

Now an error will be shown if the `pull new` or `pull attach` commands
have no remote tracking branch or a `hub.pullbase` configuration is
present.

Fixes #174.

llucax commented:
I just attached a fix. I just removed the `master` default completely. There is already a way to opt-in for a default (by configuring `hub.pullbase`, so there is not much else that needs to be done and I consider, by having a way to enable the old default, a not-so-disrupting change enough to go in the next minor release instead of a new major.

If you can give it a try let me know how it went.

llucax commented:
@joshtriplett if you want to give it a quick try (or anyone else), you can just download: https://raw.githubusercontent.com/llucax/git-hub/no-master-base-default/git-hub and run it using python.

For example:
```sh
curl https://raw.githubusercontent.com/llucax/git-hub/no-master-base-default/git-hub > /tmp/git-hub
python3 /tmp/git-hub pull new ...
```

llucax commented:
@joshtriplett @jwilk if I don't get any feedback soon I will merge this soon assuming is good enough to close this issue.

llucax commented:
Right! Thanks, I completely forgot about this one!
